lib/class-wp-theme-json-gutenberg.php

```diff
@@ -2666,6 +2666,30 @@ static function ( $pseudo_selector ) use ( $selector ) {
 			$block_rules .= static::to_ruleset( $style_variation_selector, $individual_style_variation_declarations );
 		}
 
+		// 7. Generate :root level rules related to background.
+		// Abstract into something like static::update_separator_declarations.
+		// @TODO - how can this be overwritten by theme.json, if at all? styles.dimensions.minHeight?
+		// @TODO - implement the same logic in the editor https://github.com/WordPress/gutenberg/blob/f079bd2f7fe8e4c5694fc3feb276567777f9997b/packages/block-editor/src/components/iframe/index.js#L252-L252
+		if ( static::ROOT_BLOCK_SELECTOR === $selector ) {
```

Comment:
@andrewserong @tellthemachines 

Before I go too deep on this, I wanted to garner a few opinions on how we might do this.

The PR's [description](https://github.com/WordPress/gutenberg/pull/59809#issue-2183084250) outlines the why of this PR, but the technical criteria is that, when there's a gradient background or background size of cover at the root level, we want to force the window (`HTML` or `:root`) to have a `min-height: 100%`.

This is a naive attempt to add the requisite CSS.

The first, hopeful, attempt was to switch the `static::ROOT_BLOCK_SELECTOR` from `'body'` to `:root/html`, and then set the value of `styles.dimensions.minHeight`. This would have been my preferred approach, however, it came with a bunch of unknowns, one of the weirdest being that it doesn't allow themes to overwrite a [browser's default margins](https://www.miriamsuzanne.com/2022/07/04/body-margin-8px/).

Then I mused whether `html` could be a top level "element", and therefore be "styleable", but I abandoned that idea. It might have legs, I don't know, but it'd be a bunch of work.

Then I landed here, at the MVP. 

It works, but I would be glad of alternative points of view 😄 

Some considerations:

- How, if at all, could a theme developer overwrite the min-height value? Should they at all? Maybe not since it's tightly coupled to the background option. Or am I thinking to hard about it? 😄 
- I plan to abstract the logic - if you can see avenues for optimization, that'd be great!

Reply:
Just some quick thoughts off the top of my head:

* I think it's fine if the output is `:root` but that the `ROOT_BLOCK_SELECTOR` is still `body` as the rules here can be different to the selector since it's a slightly non-standard use case. The important thing IMO is that we're outputting root rules of some kind just a single time. Out of curiosity, why do we need the rule to be attached to `:root` instead of `body`? Is it because of potential margins on `body`?
* Should there be a way for folks to explicitly switch this setting off? I do like that it's guarded behind only being output when `background` or `backgroundImage` is set at the root level 👍.
* Since you asked the question of whether a theme developer could overwrite the min-height value, I think I can imagine someone wanting to do just that! I don't think it'd necessarily need to be implemented in this PR, but if a theme were to set the root `styles.dimensions.minHeight` value in their `theme.json` file, perhaps that could overwrite the value used here? In that case, though, I imagine the theme developers would expect the value to always be output, not just when `background` is set 🤔
* In terms of abstraction, what did you have in mind? If it's only being used in one place, I like the idea of it essentially being private code so that it could be easier to change in follow-ups, potentially.

Thanks for digging in here!

Reply:
> How, if at all, could a theme developer overwrite the min-height value? Should they at all?

Is there a need for overriding this value? It's an essential part of making sure gradients and background images work correctly on the site background. 

A possible alternative would be setting min-height to a custom property via the background block support (so we'd only output it if there were a background gradient/image), and then we'd only have to set the custom property's value in this function when `static::ROOT_BLOCK_SELECTOR === $selector`. However, that would mean outputting `min-height: var(--something)` whenever there were a background set. Would we ever want to set min-height together with background in other circumstances?

Reply:
On the topic of abstraction, I'm not entirely sure about the reasoning behind `static::update_separator_declarations` but it looks like something that should be solved in the block itself by skipping serialization and adding custom logic to deal with it 🤔  It would be great to _reduce_ the amount of functions in the theme.json class if at all possible.

Reply:
Thanks a lot for the helpful feedback! I feel like I have some paths forward, now 🙇🏻 

> Out of curiosity, why do we need the rule to be attached to :root instead of body? Is it because of potential margins on body?

Good question! Two reasons:

1. `min-height: 100%` on the body tag doesn't have any effect on when using background gradients. `min-height: 100vh`, however, does. See: https://codepen.io/BoganJustice/pen/jORMKKE
2. I couldn't quite get the scrolling to behave with adding it to the body tag because of various margins of nested elements. The video below is testing `100vh` but it's also the same for `100%`


https://github.com/WordPress/gutenberg/assets/6458278/35b4bd34-b2e8-43b1-9182-40facb8b72dd


I haven't reached the end of experimentation, yet though. If it's just the logged-in admin view, then it might be something we can deal with.

Also, I suspect `min-height: 100vh;` on the body takes care of most situations, in which case it would fit nicely with theme.json `styles.dimensions.minHeight`.

> Should there be a way for folks to explicitly switch this setting off? I do like that it's guarded behind only being output when background or backgroundImage is set at the root level

If I can get it to play with `styles.dimensions.minHeight` then maybe it'll be already taken care of.

> Is there a need for overriding this value? It's an essential part of making sure gradients and background images work correctly on the site background.

You're right. It must be `100%` for it to have the intended effect on the background. I like the idea of creating a CSS custom property to be more flexible - that could potentially be overridable. 

But yeah, it might be good enough to start with the CSS rule and then decide before creating a global CSS property that folks might rely on.

> If it's only being used in one place, I like the idea of it essentially being private code so that it could be easier to change in follow-ups, potentially
> It would be great to reduce the amount of functions in the theme.json class if at all possible.

Good points. I'll leave the code in place for now. I'm not excited about another for loop there, but I can't see a tidy alternative.

Reply:
> I like the idea of creating a CSS custom property to be more flexible

The custom property wasn't for flexibility, it was a way to avoid the for loop here 😅 .
The idea is you'd set `min-height: var(--something)` in the background block support, whenever you're setting gradients or background images. But the property itself isn't set _unless_ the declarations are being applied to the root level. You'd probably need something like `html { --something: 100%; } body { --something: unset } so that value doesn't accidentally apply to all background supports. It's messy though, unless there's a use for min-height in non-root cases.